Nexus: Minimal rework for relative imports/packaging

brockdyer03 opened the pull request:
## Proposed changes

This changes the directory structure of Nexus and changes the imports such that they are relative to the base package. For the most part this PR only changes the imports and the environment requirements for testing to reflect the new directory structure.

## What type(s) of changes does this code introduce?

- Other: Restructuring and refactoring

### Does this introduce a breaking change?

- Yes

## What systems has this change been tested on?

@jtkrogel's laptop.

## Checklist

* * [ ] I have read the pull request guidance and develop docs
* * [ ] This PR is up to date with the current state of 'develop'
* * [ ] Code added or changed in the PR has been clang-formatted
* * [ ] This PR adds tests to cover any new code, or to catch a bug that is being fixed
* * [ ] Documentation has been added (if appropriate)

brockdyer03 committed:
[Nexus] Initial attempt to refactor

brockdyer03 committed:
[Nexus] Fix tests (only for `test_structure.py`)

brockdyer03 committed:
[Nexus] Tentative restructure, needs more work.

brockdyer03 committed:
[Nexus] Isolate early tests from later tests

brockdyer03 committed:
[Nexus] Remove output folders

Need to add new directory for tests to `.gitignore` at some point

jtkrogel committed:
nexus: fix visible import errors

brockdyer03 committed:
Merge pull request #1 from jtkrogel/relative_imports_minimal

[Nexus] Fix visible import errors

brockdyer03 committed:
[Nexus] Re-add `.readthedocs.yaml`

brockdyer03 committed:
[Nexus] Update `.gitignore` to reflect new test outputs, and fix `CMakeLists.txt` to use new Nexus test locations

brockdyer03 committed:
[Nexus] Fix installer to reflect new directory structure.

jtkrogel committed:
nexus: save users from themselves

brockdyer03 committed:
Merge pull request #2 from jtkrogel/relative_imports_minimal

[Nexus] Save users from themselves

brockdyer03 committed:
[Nexus] Add explicit path to correct Nexus package for user examples.

brockdyer03 committed:
[Nexus] Try alternate way to protect users from themselves

brockdyer03 committed:
[Nexus] Fix bin tests to use proper `PYTHONPATH`

Also removed unused import `divert_nexus` from `test_nxs_sim`, and made sure to add `import os` in `qdens-radial`

prckent commented on nexus/bin/nxs-test:
```diff
@@ -1797,7 +1796,7 @@ def user_examples(label):
     for script in einfo['scripts']:
         # run the example script
-        command = sys.executable+' ./'+script+' --generate_only --sleep=0.01'
+        command = 'PYTHONPATH="{}" '.format(sys.path[0])+sys.executable+' ./'+script+' --generate_only --sleep=0.01'
```
Why is this PYTHONPATH setting now required?

brockdyer03 committed:
Revert "[Nexus] Fix bin tests to use proper `PYTHONPATH`"

This reverts commit c9d965c1d12612dc5f5396949ddfc157adf88eb2.

brockdyer03 committed:
Revert "[Nexus] Add explicit path to correct Nexus package for user examples."

This reverts commit 8c1a3eea514bb76b91c64ec1dc8b3f9ca335cc88.

brockdyer03 committed:
[Nexus] Revert Python pathing changes and alter `CMakeLists.txt` instead

prckent commented:
Is this still WIP/Draft or ready for final review and merge?

[ Give the many files that are updated, it would be ideal to get this merged promptly if possible to keep merges simple for other developers ]

prckent reviewed (changes requested):
All references to nexus/lib in the repo need updating to nexus/nexus (grep -R -n -i nexus/lib *)
```
docs/examples.rst:39:For Python to find the Nexus library, the PYTHONPATH environment variable should be set to ``<QMCPACK source>/nexus/lib``.
docs/introduction.rst:94:   export PYTHONPATH=$PYTHONPATH:location-of-qmcpack/nexus/library
nexus/docs/installation.rst:20:  export PYTHONPATH=/your_download_path/nexus/lib:$PYTHONPATH
nexus/docs/installation.rst:289:  >coverage report | grep nexus/lib
nexus/docs/installation.rst:291:  nexus/lib/basisset.py                      631    375    41%
nexus/docs/installation.rst:292:  nexus/lib/bundle.py                        191     68    64%
nexus/docs/installation.rst:293:  nexus/lib/debug.py                          12      6    50%
nexus/docs/installation.rst:294:  nexus/lib/developer.py                     261     97    63%
nexus/docs/installation.rst:295:  nexus/lib/execute.py                        13      2    85%
nexus/docs/installation.rst:296:  nexus/lib/fileio.py                        957    373    61%
nexus/docs/installation.rst:297:  nexus/lib/gamess.py                        102     20    80%
nexus/docs/installation.rst:298:  nexus/lib/gamess_analyzer.py               305    149    51%
nexus/docs/installation.rst:299:  nexus/lib/gamess_input.py                  597    167    72%
nexus/docs/installation.rst:300:  nexus/lib/generic.py                       817    173    79%
nexus/docs/installation.rst:301:  nexus/lib/grid_functions.py               1192    435    64%
nexus/docs/installation.rst:302:  nexus/lib/hdfreader.py                     215     61    72%
nexus/docs/installation.rst:303:  nexus/lib/machines.py                     1887    463    75%
nexus/docs/installation.rst:304:  nexus/lib/memory.py                         60      7    88%
nexus/docs/installation.rst:305:  nexus/lib/nexus.py                         297    140    53%
nexus/docs/installation.rst:306:  nexus/lib/nexus_base.py                     74     11    85%
nexus/docs/installation.rst:307:  nexus/lib/numerics.py                      756    372    51%
nexus/docs/installation.rst:308:  nexus/lib/periodic_table.py               1505     24    98%
nexus/docs/installation.rst:309:  nexus/lib/physical_system.py               427     73    83%
nexus/docs/installation.rst:310:  nexus/lib/plotting.py                       22      7    68%
nexus/docs/installation.rst:311:  nexus/lib/project_manager.py               234     37    84%
nexus/docs/installation.rst:312:  nexus/lib/pseudopotential.py              1225    559    54%
nexus/docs/installation.rst:313:  nexus/lib/pwscf.py                         198     73    63%
nexus/docs/installation.rst:314:  nexus/lib/pwscf_analyzer.py                634    316    50%
nexus/docs/installation.rst:315:  nexus/lib/pwscf_data_reader.py             132    120     9%
nexus/docs/installation.rst:316:  nexus/lib/pwscf_input.py                  1261    563    55%
nexus/docs/installation.rst:317:  nexus/lib/pwscf_postprocessors.py          434     56    87%
nexus/docs/installation.rst:318:  nexus/lib/pyscf_analyzer.py                  3      0   100%
nexus/docs/installation.rst:319:  nexus/lib/pyscf_input.py                   181     26    86%
nexus/docs/installation.rst:320:  nexus/lib/pyscf_sim.py                      57      8    86%
nexus/docs/installation.rst:321:  nexus/lib/qmcpack.py                       344    146    58%
nexus/docs/installation.rst:322:  nexus/lib/qmcpack_analyzer.py              457    104    77%
nexus/docs/installation.rst:323:  nexus/lib/qmcpack_analyzer_base.py         327    137    58%
nexus/docs/installation.rst:324:  nexus/lib/qmcpack_converters.py            507     83    84%
nexus/docs/installation.rst:325:  nexus/lib/qmcpack_input.py                3605   1439    60%
nexus/docs/installation.rst:326:  nexus/lib/qmcpack_method_analyzers.py      198     64    68%
nexus/docs/installation.rst:327:  nexus/lib/qmcpack_property_analyzers.py    205     97    53%
nexus/docs/installation.rst:328:  nexus/lib/qmcpack_quantity_analyzers.py   2070   1789    14%
nexus/docs/installation.rst:329:  nexus/lib/qmcpack_result_analyzers.py      285    142    50%
nexus/docs/installation.rst:330:  nexus/lib/quantum_package.py               253    141    44%
nexus/docs/installation.rst:331:  nexus/lib/quantum_package_analyzer.py        3      0   100%
nexus/docs/installation.rst:332:  nexus/lib/quantum_package_input.py         338    164    51%
nexus/docs/installation.rst:333:  nexus/lib/simulation.py                   1019    169    83%
nexus/docs/installation.rst:334:  nexus/lib/structure.py                    3830   2055    46%
nexus/docs/installation.rst:335:  nexus/lib/superstring.py                   311    199    36%
nexus/docs/installation.rst:336:  nexus/lib/testing.py                       409     67    84%
nexus/docs/installation.rst:337:  nexus/lib/unit_converter.py                121      4    97%
nexus/docs/installation.rst:338:  nexus/lib/vasp.py                           94     15    84%
nexus/docs/installation.rst:339:  nexus/lib/vasp_analyzer.py                 548     73    87%
nexus/docs/installation.rst:340:  nexus/lib/vasp_input.py                    906    412    55%
nexus/docs/installation.rst:341:  nexus/lib/versions.py                      335     50    85%
nexus/docs/installation.rst:342:  nexus/lib/xmlreader.py                     260     54    79%
nexus/README:76:     export PYTHONPATH=/your/path/to/nexus/lib/:$PYTHONPATH
tests/test_automation/nightly_test_scripts/nightly_ornl.sh:219:export PYTHONPATH=${test_dir}/qmcpack/nexus/lib
tests/test_automation/containers/Dockerfile_gcc_mpi_develop_complete:81:ENV PYTHONPATH=/opt/qmcpack/qmcpack/nexus/lib
utils/convert-gamess-ecp-to-qmcpack:4:nexus_lib = os.path.join(os.path.dirname(__file__), '../nexus/lib')
```

Please also verify that all the executables in nexus/bin run correctly. My 1st attempts have problems, but I could have a bad path etc.

brockdyer03 committed:
[Nexus] Replace mentions of `nexus/lib` with updated references.

I also went and updated every import of a Nexus module in QMCPACK, e.g. `grep -R "from <nexus module>" *.py`, and additionally fixed the Dockerfile so that it exports the correct `$PYTHONPATH`.

brockdyer03 committed:
Merge branch 'develop' into relative_imports_minimal

prckent reviewed (dismissed):
LGTM

FYI: I checked the current coverage statistics:

```
qmcuser@ec42f35e876e:~/qmcpack/nexus/bin$ python3-coverage report --include="/home/qmcuser/qmcpack/nexus/nexus/*" --omit="/home/qmcuser/qmcpack/nexus/nexus/tests/*"
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py:112: SyntaxWarning: invalid escape sequence '\p'
  """
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py:156: SyntaxWarning: invalid escape sequence '\m'
  """
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py:200: SyntaxWarning: invalid escape sequence '\p'
  """
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py:250: SyntaxWarning: invalid escape sequence '\m'
  """
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py:2395: SyntaxWarning: invalid escape sequence '\p'
  """
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py:2770: SyntaxWarning: invalid escape sequence '\p'
  """
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_analyzer.py:186: SyntaxWarning: invalid escape sequence '\ '
  """
/home/qmcuser/qmcpack/nexus/nexus/structure.py:25: SyntaxWarning: invalid escape sequence '\_'
  """
/home/qmcuser/qmcpack/nexus/nexus/xmlreader.py:410: SyntaxWarning: invalid escape sequence '\d'
  epattern = re.compile(name+'\d+')
Name                                                              Stmts   Miss  Cover
-------------------------------------------------------------------------------------
/home/qmcuser/qmcpack/nexus/nexus/__init__.py                       305    141    54%
/home/qmcuser/qmcpack/nexus/nexus/basisset.py                       643    387    40%
/home/qmcuser/qmcpack/nexus/nexus/bundle.py                         191     68    64%
/home/qmcuser/qmcpack/nexus/nexus/debug.py                           12      6    50%
/home/qmcuser/qmcpack/nexus/nexus/developer.py                      265     88    67%
/home/qmcuser/qmcpack/nexus/nexus/execute.py                         14      2    86%
/home/qmcuser/qmcpack/nexus/nexus/fileio.py                        1016    423    58%
/home/qmcuser/qmcpack/nexus/nexus/gamess.py                         151     62    59%
/home/qmcuser/qmcpack/nexus/nexus/gamess_analyzer.py                306    131    57%
/home/qmcuser/qmcpack/nexus/nexus/gamess_input.py                   593    167    72%
/home/qmcuser/qmcpack/nexus/nexus/generic.py                        832    179    78%
/home/qmcuser/qmcpack/nexus/nexus/grid_functions.py                1641    787    52%
/home/qmcuser/qmcpack/nexus/nexus/hdfreader.py                      205     59    71%
/home/qmcuser/qmcpack/nexus/nexus/machines.py                      2500    671    73%
/home/qmcuser/qmcpack/nexus/nexus/memory.py                          60      7    88%
/home/qmcuser/qmcpack/nexus/nexus/nexus_base.py                      72      9    88%
/home/qmcuser/qmcpack/nexus/nexus/nexus_version.py                    2      0   100%
/home/qmcuser/qmcpack/nexus/nexus/numerics.py                       903    500    45%
/home/qmcuser/qmcpack/nexus/nexus/observables.py                    890    615    31%
/home/qmcuser/qmcpack/nexus/nexus/periodic_table.py                1504     24    98%
/home/qmcuser/qmcpack/nexus/nexus/physical_system.py                433     72    83%
/home/qmcuser/qmcpack/nexus/nexus/project_manager.py                233     37    84%
/home/qmcuser/qmcpack/nexus/nexus/pseudopotential.py               1673   1002    40%
/home/qmcuser/qmcpack/nexus/nexus/pwscf.py                          216     87    60%
/home/qmcuser/qmcpack/nexus/nexus/pwscf_analyzer.py                 653    283    57%
/home/qmcuser/qmcpack/nexus/nexus/pwscf_data_reader.py              129    120     7%
/home/qmcuser/qmcpack/nexus/nexus/pwscf_input.py                   1386    671    52%
/home/qmcuser/qmcpack/nexus/nexus/pwscf_postprocessors.py           517    113    78%
/home/qmcuser/qmcpack/nexus/nexus/pyscf_analyzer.py                   3      0   100%
/home/qmcuser/qmcpack/nexus/nexus/pyscf_input.py                    285     44    85%
/home/qmcuser/qmcpack/nexus/nexus/pyscf_sim.py                       64     14    78%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack.py                       1037    787    24%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_analyzer.py               456    106    77%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_analyzer_base.py          326    136    58%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_converters.py             673    214    68%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_input.py                 4421   1988    55%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_method_analyzers.py       196     64    67%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_property_analyzers.py     203    100    51%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_quantity_analyzers.py    2094   1818    13%
/home/qmcuser/qmcpack/nexus/nexus/qmcpack_result_analyzers.py       289    144    50%
/home/qmcuser/qmcpack/nexus/nexus/quantum_package.py                253    141    44%
/home/qmcuser/qmcpack/nexus/nexus/quantum_package_analyzer.py         3      0   100%
/home/qmcuser/qmcpack/nexus/nexus/quantum_package_input.py          337    164    51%
/home/qmcuser/qmcpack/nexus/nexus/rmg.py                             33     10    70%
/home/qmcuser/qmcpack/nexus/nexus/rmg_analyzer.py                   293    265    10%
/home/qmcuser/qmcpack/nexus/nexus/rmg_input.py                      681    163    76%
/home/qmcuser/qmcpack/nexus/nexus/simulation.py                    1055    214    80%
/home/qmcuser/qmcpack/nexus/nexus/structure.py                     4157   2485    40%
/home/qmcuser/qmcpack/nexus/nexus/testing.py                        448     88    80%
/home/qmcuser/qmcpack/nexus/nexus/unit_converter.py                 120      2    98%
/home/qmcuser/qmcpack/nexus/nexus/utilities.py                       56     16    71%
/home/qmcuser/qmcpack/nexus/nexus/vasp.py                            93     15    84%
/home/qmcuser/qmcpack/nexus/nexus/vasp_analyzer.py                  546     73    87%
/home/qmcuser/qmcpack/nexus/nexus/vasp_input.py                     956    454    53%
/home/qmcuser/qmcpack/nexus/nexus/versions.py                       349     44    87%
/home/qmcuser/qmcpack/nexus/nexus/xmlreader.py                      298     53    82%
-------------------------------------------------------------------------------------
TOTAL                                                             37070  16313    56%
```

brockdyer03 committed:
[Nexus] Update `CHANGELOG.md` to reflect recent changes.

prckent committed:
Refine CHANGELOG.md

ye-luo committed:
Merge branch 'develop' into relative_imports_minimal
